File: medfuncSLfilteredbackprojection.cpp
#include<vector>
#include <cmath>
std::vector<std::vector<double>> medfuncSLfilteredbackprojection(int theta_num, int N, std::vector<std::vector<double>> R, double delta, std::vector<double>fh_SL) {
	std::vector<std::vector<double>> rec_SL(N, std::vector<double>(N, 0));
	for (int i = 0; i < theta_num; i++) {
		std::vector<double> pm(N, 0);
		for (int j = 0; j < N; j++) {
			pm[j] = R[j][i];
		}
		int lenA = fh_SL.size();
		int lenB = pm.size();
		int lenOut = lenA + lenB - 1;
		std::vector<double> conv(lenOut, 0);
		// 执行线性卷积
		for (int m = 0; m < lenOut; ++m) {
			for (size_t n = 0; n <= m; ++n) {
				if (n < lenA && m - n < lenB) {
					conv[m] += fh_SL[n] * pm[m - n];
				}
			}
		}
		std::vector<double> pm_SL(N, 0);
		for (int m = 0; m < N; m++) {
			pm_SL[m] = conv[ceil((lenOut - lenA) / 2) - 1 + m];
		}
		double Cm = (N / 2) * (1 - cos(i * delta) - sin(i * delta));
		for (int j = 0; j < N; j++) {
			for (int k = 0; k < N; k++) {
				double Xrm = Cm + k * cos(i * delta) + j * sin(i * delta);
				int n = floor(Xrm);
				double t = Xrm - n;
				if (n < 0) n = 0;
				if (n > N - 2)n = N - 2;
				double p = (1 - t) * pm_SL[n] + t * pm_SL[n + 1];
				rec_SL[k][j] = rec_SL[k][j] + p;
			}
		}
	}
	return rec_SL;
}

File: medfuncSLfilteredbackprojection.m
function rec_SL=medfuncSLfilteredbackprojection(theta_num,N,R1,delta,fh_SL)
rec_SL=zeros(N);
for m=1:theta_num
    pm=R1(:,m);
    pm_SL=conv(fh_SL,pm,'same');
    Cm=(N/2)*(1-cos((m-1)*delta)-sin((m-1)*delta));
    for k1=1:N
        for k2=1:N
            Xrm=Cm+(k2-1)*cos((m-1)*delta)+(k1-1)*sin((m-1)*delta);
            n=floor(Xrm);
            t=Xrm-floor(Xrm);
            n=max(1,n);n=min(n,N-1);
            p_SL=(1-t)*pm_SL(n)+t*pm_SL(n+1);
            rec_SL(N+1-k1,k2)=rec_SL(N+1-k1,k2)+p_SL;
        end
    end
end
end